receiver/collectdreceiver/collectd.go

```diff
@@ -0,0 +1,292 @@
+	}
+	return metricGauge(isDouble)
+}
+
+func (r *collectDRecord) newPoint(val *json.Number) (*metricspb.Point, bool, error) {
+	p := &metricspb.Point{
+		Timestamp: r.protoTime(),
+	}
+
+	isDouble := true
+	if v, err := val.Int64(); err == nil {
+		isDouble = false
+		p.Value = &metricspb.Point_Int64Value{Int64Value: v}
+	} else {
+		v, err := val.Float64()
+		if err != nil {
+			return nil, isDouble, fmt.Errorf("value could not be decoded: %v", err)
+		}
+		p.Value = &metricspb.Point_DoubleValue{DoubleValue: v}
+	}
+	return p, isDouble, nil
+}
+
+// getReasonableMetricName creates metrics names by joining them (if non empty) type.typeinstance
+// if there are more than one dsname append .dsname for the particular uint. if there's only one it
+// becomes a dimension.
+func (r *collectDRecord) getReasonableMetricName(index int, attrs map[string]string) (string, bool) {
+	usedDsName := false
+	parts := make([]byte, 0, len(*r.TypeS)+len(*r.TypeInstance))
+
+	if !isNilOrEmpty(r.TypeS) {
+		parts = append(parts, *r.TypeS...)
+	}
+	parts = r.pointTypeInstance(attrs, parts)
+	if r.Dsnames != nil && !isNilOrEmpty(r.Dsnames[index]) && len(r.Dsnames) > 1 {
+		if len(parts) > 0 {
+			parts = append(parts, '.')
+		}
+		parts = append(parts, *r.Dsnames[index]...)
+		usedDsName = true
+	}
+	return string(parts), usedDsName
+}
+
+// pointTypeInstance extracts information from the TypeInstance field and appends to the metric name when possible.
+func (r *collectDRecord) pointTypeInstance(attrs map[string]string, parts []byte) []byte {
+	if !isNilOrEmpty(r.TypeInstance) {
+		instanceName, extractedAttrs := labelsFromName(r.TypeInstance)
+		if instanceName != "" {
+			if len(parts) > 0 {
+				parts = append(parts, '.')
+			}
+			parts = append(parts, instanceName...)
+		}
+		for k, v := range extractedAttrs {
+			if _, exists := attrs[k]; !exists {
+				val := v
+				addIfNotNullOrEmpty(attrs, k, true, &val)
+			}
+		}
+	}
+	return parts
+}
+
+// labelsFromName tries to pull out dimensions out of name in the format name[k=v,f=x]-morename
+// would return name-morename and extract dimensions (k,v) and (f,x)
+// if we encounter something we don't expect use original
+// this is a bit complicated to avoid allocations, string.split allocates, while slices
+// inside same function, do not.
+func labelsFromName(val *string) (instanceName string, toAddDims map[string]string) {
+	instanceName = *val
+	index := strings.Index(*val, "[")
+	if index > -1 {
+		left := (*val)[:index]
+		rest := (*val)[index+1:]
+		index = strings.Index(rest, "]")
+		if index > -1 {
+			working := make(map[string]string)
+			dimensions := rest[:index]
+			rest = rest[index+1:]
+			cindex := strings.Index(dimensions, ",")
+			prev := 0
+			for {
+				if cindex < prev {
+					cindex = len(dimensions)
+				}
+				piece := dimensions[prev:cindex]
+				tindex := strings.Index(piece, "=")
+				//lint:ignore S1003 staticcheck wants us to use strings.Conains instead of Index
```

Comment:
We use the value of tindex later in the function and on the next line as well so we need to use Index. I think staticcheck wrongly assumes we are only checking for existence of the char at index because `if index == -1` following the .Index() call.

Reply:
I was think that the `lint:ignore` was for the one on the if statement for which the code just compares the return value.